netket/_driver.py

```diff
@@ -0,0 +1,265 @@
+            machine: The variational wavefunction.
+            sampler: The Monte Carlo sampler.
+            optimizer: Determines how optimization steps are performed given the
+                bare energy gradient. This parameter supports three different kinds of inputs,
+                which are described in the docs of `make_optimizer_fn`.
+            n_samples: Number of Markov Chain Monte Carlo sweeps to be
+                performed at each step of the optimization.
+            discarded_samples (int): Number of sweeps to be discarded at the
+                beginning of the sampling, at each step of the optimization.
+                Defaults to 10% of n_samples.
+            discarded_samples_on_init (int): Number of sweeps to be discarded in
+                the first step of optimization, at the beginning of the
+                sampling.
+            sr (SR, optional): Determines whether and how stochastic reconfiguration
+                is applied to the bare energy gradient before performing applying
+                the optimizer. If this parameter is not passed or None, SR is not used.
+
+        Example:
+            Optimizing a 1D wavefunction with Variational Mante Carlo.
+
+            ```python
+            >>> import netket as nk
+            >>> SEED = 3141592
+            >>> g = nk.graph.Hypercube(length=8, n_dim=1)
+            >>> hi = nk.hilbert.Spin(s=0.5, graph=g)
+            >>> ma = nk.machine.RbmSpin(hilbert=hi, alpha=1)
+            >>> ma.init_random_parameters(seed=SEED, sigma=0.01)
+            >>> ha = nk.operator.Ising(hi, h=1.0)
+            >>> sa = nk.sampler.MetropolisLocal(machine=ma)
+            >>> sa.seed(SEED)
+            >>> op = nk.optimizer.Sgd(learning_rate=0.1)
+            >>> vmc = nk.variational.Vmc(hamiltonian=ha, machine=ma, sampler=sa,
+            ... optimizer=op, n_samples=500)
+            >>> print(ma.n_visible)
+            8
+            ```
+        """
+        self._ham = hamiltonian
+        self._machine = machine
+        self._sampler = sampler
+        self._sr = sr
+
+        self._optimizer_step, self._optimizer_desc = make_optimizer_fn(optimizer)
+
+        self._npar = machine.n_par
+        self._mc_data = None
+
+        self.n_samples = n_samples
+        self.n_discard = n_discard if n_discard else n_samples // 10
+
+        self._obs = {}
+
+        self.step_count = 0
+
+    def _get_mc_stats(self, op):
+        loc = _local_values(
+            self._mc_data.samples, self._mc_data.log_values, self._machine, op
+        )
+        return loc, _statistics(loc)
+
+    def advance(self, n_steps=1):
+        """
+        Performs a number of VMC optimization steps.
+
+        Args:
+            n_steps (int): Number of steps to perform.
+        """
+
+        def update_samples():
+            self._mc_data = _vmc.compute_samples(
+                self._sampler, self.n_samples, self.n_discard, der_logs="centered"
```

Comment:
Yes, I really think we should remove der_logs from here! It would be much less error prone etc

Reply:
If having two separate calls does not impact performance, that would be the best solution, IMHO. Then one could call
```python
samples, logpsi = _vmc.compute_samples(...) # wrapping samples and logpsi into one object should not be needed anymore, so just return a tuple
logder = _vmc.log_derivatives(samples, centered=True)

eloc = _vmc.local_values(samples, logpsi, self.ham, self.machine)
grad_E = _vmc.gradient_of_expectation(eloc, logder)
```